src/stream/src/executor/backfill/cdc/cdc_backfill.rs

```diff
@@ -168,29 +175,107 @@ impl<S: StateStore> CdcBackfillExecutor<S> {
 
-        let mut upstream = transform_upstream(upstream, &self.output_columns)
-            .boxed()
-            .peekable();
-
         state_impl.init_epoch(first_barrier_epoch).await?;
 
         // restore backfill state
         let state = state_impl.restore_state().await?;
         current_pk_pos = state.current_pk_pos.clone();
 
-        let to_backfill = !self.options.disable_backfill && !state.is_finished;
+        let need_backfill = !self.options.disable_backfill && !state.is_finished;
 
         // Keep track of rows from the snapshot.
         let mut total_snapshot_row_count = state.row_count as u64;
 
+        // After init the state table and forward the initial barrier to downstream,
+        // we now try to create the table reader with retry.
+        // If backfill hasn't finished, we can ignore upstream cdc events before we create the table reader;
+        // If backfill is finished, we should forward the upstream cdc events to downstream.
+        let mut upstream_buffer = UpstreamMessageBuffer::new(upstream);
+        let table_reader: Option<ExternalTableReaderImpl>;
+        let mut backoff = get_backoff_strategy();
+        loop {
+            let create_result: StreamExecutorResult<_> = try {
+                let reader = upstream_buffer
+                    .run_future(create_table_reader(&self.external_table))
+                    .await?;
+                reader
+            };
+
+            // emit upstream message if any
+            if let Some(msg) = upstream_buffer.next_pending_message().await? {
+                match msg {
+                    Message::Barrier(barrier) => {
+                        // commit state to bump the epoch of state table
+                        state_impl.commit_state(barrier.epoch).await?;
+                        yield Message::Barrier(barrier);
+                    }
+                    Message::Chunk(chunk) => {
+                        if need_backfill {
+                            // ignore chunk if we need backfill, since we can read the data from the snapshot
+                        } else {
+                            // forward the chunk to downstream
+                            yield Message::Chunk(chunk);
+                        }
+                    }
+                    Message::Watermark(_) => {
+                        // ignore watermark
+                    }
+                }
+            }
+
+            match create_result {
+                Ok(reader) => {
+                    table_reader = Some(reader);
+                    break;
+                }
+                Err(e) => {
+                    tracing::warn!(
+                        table_id,
+                        upstream_table_name,
+                        error = %e.as_report(),
+                        "failed to create external table reader",
+                    );
+                    if let Some(delay) = backoff.next() {
+                        sleep(delay).await;
```

Comment:
This backoff will also cause delay in polling the upstream buffer so there is still side effect here when the upstream DB is unavailable.